src/proposal.jl

```diff
@@ -26,29 +34,83 @@ function SliceNUTS(h::Hamiltonian, θ::AbstractVector{T}) where {T<:Real}
     return SliceNUTS(Leapfrog(find_good_eps(h, θ)))
 end
 
+function find_good_eps(rng::AbstractRNG, h::Hamiltonian, θ::AbstractVector{T}; max_n_iters::Int=100) where {T<:Real}
```

Comment:
Consider moving

```julia
    r = rand_momentum(rng, h)
    H = hamiltonian_energy(h, θ, r)
```

out of `find_good_eps`  and wraping


```julia
        θ′, r′, _is_valid = step(Leapfrog(ϵ′), h, θ′, r′)
        H_new = _is_valid ? hamiltonian_energy(h, θ′, r′) : Inf
```

into a  function 

```julia

function A(ϵ, h, θ, r) 
        θ′, r′, _is_valid = step(Leapfrog(ϵ′), h, θ′, r′)
        return H_new = _is_valid ? hamiltonian_energy(h, θ′, r′) : Inf
end
```

Then, moving this function `find_good_eps(h::Hamiltonian, θ::AbstractVector{T}, A::Function; max_n_iters::Int=100)`  into `adaption/stepsize.jl`.

Reply:
Sounds good. Would it be reusable for MH then?

Reply:
I'm going to do this in another PR as a generalised implementation (which can also be used for MH) requires to resolve https://github.com/TuringLang/AdvancedHMC.jl/issues/17. Why? Because we need a unified type for MH as well, which doesn't have momentum.